### DISCUSSION TITLE
变量作用域

### DISCUSSION THREAD

**Discussion by daiwhea** [zigcc/forum#95] [新手提问] [answered]
```
std.io.Reader 中的代码
pub fn readStruct(self: Self, comptime T: type) anyerror!T {
    // Only extern and packed structs have defined in-memory layout.
    comptime assert(@typeInfo(T).Struct.layout != .Auto);
    var res: [1]T = undefined;  // A
    try self.readNoEof(mem.sliceAsBytes(res[0..]));
    return res[0];  // B
}
```

1. 这段代码中，A定义的 [1]T 是哪里创建的？因为在 File.reader() 以及 readStruct() 里面都没有传入 allocator ，A这里定义的变量，却可以在 B 返回给调用者，并且没有被销毁。

2. A 为什么定义为 [1]T，而不是 T，很明显，这里返回的是 T，那么为啥要如 A 那样定义？

谢谢你的时间。。

**Comment by hshq** [comment, marked as answer]
1. res 在栈中，这符合没有隐式内存分配的原则；也因此返回后自然会无效；
    返回的是 T 类型，不是 *T ，T 的内容整个拷贝返回了，销毁不存在问题。

2. sliceAsBytes 需要切片，所以用 [1]T 而不是 T 是很自然的做法吧。

**Comment by daiwhea** [comment]
多谢，问题 1 是我混淆了 fn 返回的处理。
